kernel/threads.c

```diff
@@ -146,3 +146,42 @@ void block_current_thread (void) {
   running_tcb->state = TS_BLOCKED;
   schedule();
 }
+
+int fork_ret = 0;
+
+int clone_thread (uint64_t fork_rsp) {
+  /* This tells us how much of the kernel stack to copy to the new thread */
+  uint64_t stack_depth = (uint64_t) running_tcb->stack_top - fork_rsp;
+  for (int i = 0; i < NUMTHREADS; i++) {
+    if (tcbs[i].state == TS_NONEXIST) {
+      tcbs[i].thread_id = thread_id++;
+      tcbs[i].pt = new_pt();
+      clone_pagemap(&tcbs[i].pm, &running_tcb->pm);
+      tcbs[i].text = running_tcb->text;
+      tcbs[i].text_length = running_tcb->text_length;
+      /* Set up stacks */
+      char *kernel_stack = &((char *)allocate_phys_page())[4096];
+      /* Initialize tcb struct */
+      tcbs[i].state = TS_INACTIVE;
+      tcbs[i].stack_top = &kernel_stack[0];
+      memcpy(kernel_stack - stack_depth,
+          ((char *)running_tcb->stack_top) - stack_depth, stack_depth);
+      /* Initialize stack */
+      tcbs[i].rsp = kernel_stack - stack_depth;
+      tcbs[i].fpu_state = THREAD_FPU_STATE_INACTIVE;
```

Comment:
Set it to the fpu_state of the old thread; allocate and copy the fpu_buf if it's ACTIVE